File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int bs(vector<int>& nums,int start,int end, int target){
        
        int mid = 0 , result = -1;
       
        while(start <= end){
            
            mid = start + (end - start)/2;
            
            if(nums[mid] == target){
                result = mid;
                break;
            }
            else if(nums[mid] > target){
                end = mid - 1;
            }
            else{
                start = mid + 1;
            }
        }
        return result;
    }
    
    void Number_of_Rotations(vector<int> arr, int N, int &result)
    {
        int next, prev, start, end, mid;

        start = 0;
        end = N - 1;

        if(arr[start] < arr[end])
        {
            result = start;
            return;
        }
        while(start <= end)
        {

            if(arr[start] <= arr[end])
            {
                result = start;
                break;
            }
            mid = start + (end - start)/2;

            cout<<" mid : "<<mid<<endl;
            next = (mid + 1)%N;
            prev = (mid + N - 1)%N;

            cout<<" p : "<<prev<<" n : "<<next<<endl;

            if(arr[mid] <= arr[next] && arr[mid] <= arr[prev])  ///that minimum element is always in unsorted part   jo previous aur next dono se small ho
            {
                result = mid;
                break;
            }

            ///other
            if(arr[start] <= arr[mid])  ///if it is that means that array is in sorted order of that portion then
            ///Portion - 1 is Sorted                ///move to another portion and repeat process
            {
                cout<<" P - 1 "<<endl;
                start = mid + 1;
            }

            else if(arr[mid] <= arr[end])   ///Portion - 2 is Sorted
            {
                cout<<" P - 2 "<<endl;
                end = mid - 1;
            }
        }///while
    }
    
    int search(vector<int>& nums, int target) {
    
        int result = -1;

        if(nums.size() == 1 && nums[0] == target){
            return 0;    
        }else if(nums.size() == 1 && nums[0] != target){
            return -1;
        }
        
        Number_of_Rotations(nums, nums.size(), result);
        
        cout<<result<<endl;
        if(nums[result] == target){
            return result;
        }
        int x = bs(nums, 0, result, target);
        int y = bs(nums, result+1, nums.size()-1, target);
        
        if(x == -1 && y == -1){
            return -1;
        }
        if(x != -1){
            result = x;
        }
        if(y != -1){
            result = y;
        }
        return result;
    }
};

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def binary_search(self, nums: List[int], low, high, target: int) -> int:
        start = low
        end = high

        while(start <= end):
            mid = start + (end - start) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                start = mid + 1
            else:
                end = mid - 1
        return -1

    def num_rotations(self, nums, N, result):
        start = 0
        end = N - 1

        if nums[start] < nums[end]:
            result = start
            return result

        while start <= end:
            if nums[start] <= nums[end]:
                result = start
                break

            mid = start + (end - start) // 2
            nxt = (mid + 1) % N
            prev = (mid + N - 1) % N

            # that minimum element is always in unsorted part   jo previous aur next dono se small ho
            if nums[mid] <= nums[nxt] and nums[mid] <= nums[prev]:  # min element
                result = mid
                break
            # if it is that means that array is in sorted order of that portion then, move to another portion and repeat process
            if nums[start] <= nums[mid]:  # if portion 1 is sorted
                start = mid + 1
            elif nums[mid] <= nums[end]:  # if portion 2 is sorted
                end = mid - 1
        return result

    def search(self, nums: List[int], target: int) -> int:
        min_index = self.num_rotations(nums, len(nums), target)

        p1 = self.binary_search(nums, 0, min_index - 1, target)
        p2 = self.binary_search(nums, min_index, len(nums) - 1, target)

        if p1 == -1 and p2 == -1:
            return -1
        if p1 != -1:
            return p1
        if p2 != -1:
            return p2
         